pkg/table/tables/tables.go

```diff
@@ -512,20 +512,38 @@ func (t *TableCommon) UpdateRecord(ctx context.Context, sctx table.MutateContext
 		}
 	}
 
-	var colIDs, binlogColIDs []int64
-	var row, binlogOldRow, binlogNewRow []types.Datum
+	var binlogColIDs []int64
+	var binlogOldRow, binlogNewRow []types.Datum
 	var checksums []uint32
 	numColsCap := len(newData) + 1 // +1 for the extra handle column that we may need to append.
-	colIDs = make([]int64, 0, numColsCap)
-	row = make([]types.Datum, 0, numColsCap)
+
+	// a reusable buffer to save malloc
+	// Note: The buffer should not be referenced or modified outside this function.
+	// It can only act as a temporary buffer for the current function call.
+	buffer := sctx.GetSessionVars().TablesBufferPool.Update
+	if cap(buffer.ColIDs) < numColsCap {
+		buffer.ColIDs = make([]int64, 0, numColsCap)
+	} else {
+		buffer.ColIDs = buffer.ColIDs[:0]
+	}
+
+	if cap(buffer.Row) < numColsCap {
+		buffer.Row = make([]types.Datum, 0, numColsCap)
+	} else {
+		buffer.Row = buffer.Row[:0]
+	}
```

Comment:
How about extract a function for these boilerplate code. eg:
```go
func resizeSlice[T any](slice []T, size int, optCap ...int) []T {
	capacity := size
	if len(optCap) > 0 {
		capacity = optCap[0]
	}
	if cap(slice) < capacity {
		return make([]T, size, capacity)
	} else {
		return slice[:size]
	}
}

buffer.Row = resizeSlice(buffer.Row, 0, numColsCap)
```

Reply:
Good idea. We may not need the generalization of `size` and `optCap`